ENH Deprecate `class_weight_` in regression/single class models in SVM module

Micky774 opened the pull request:
#### Reference Issues/PRs
Fixes #14788

#### What does this implement/fix? Explain your changes.
Began deprecation of some properties in svm module

- Removed instation of `class_weight_ = np.empty(0)` in `BaseLibSVM`
- Added new property `_class_weight` in `BaseLibSVM` to serve as an intermediate for deprecation by avoiding direct calls to `class_weight_` for `SVR`, `NuSVR`, and `OneClassSVM`
- Added test for deprecated attribute

#### Any other comments?

Micky774 committed:
Began deprecation of some properties in svm module

- Removed instation of `class_weight_ = np.empty(0)` in `BaseLibSVM`
- Added deprecated properties for `class_weight_` and `n_support_`
  in `SVR`, `NuSVR`, and `OneClassSVM`.

Micky774 committed:
Added tests for depracation

Micky774 committed:
Added changelog entry

Micky774 committed:
Added ignore flags for mypy on deprecation wrapper

Micky774 committed:
Merge branch 'main' into deprecate_in_svr

Micky774 committed:
Added todo message for removal in 1.3

Micky774 committed:
Refactored deprecation

Micky774 committed:
Added private attribute for cleaner deprecation

thomasjpfan commented on sklearn/svm/_base.py:
This removes the error for SVR. I think we can remove the dependency on `self.n_support_`:

```python
        svm_type = LIBSVM_IMPL.index(self._impl)
        if svm_type in (0, 1):
            total_support = self._n_support.sum()
        else:
            total_support = self._n_support[0]

        if not self._sparse and sv.size > 0 and total_support != sv.shape[0]:
            raise ValueError(
                f"The internal representation of {self.__class__.__name__} was altered"
            )
```

thomasjpfan commented on sklearn/svm/_base.py:
```diff
@@ -682,10 +686,26 @@ def n_support_(self):
             return self._n_support
         else:
-            # SVR and OneClass
+            # SVR, NuSVR, and OneClass
```
I think we need to have the subclasses define the `n_support_` property so it does not show up in the documentation when we remove `n_support_`.

In the other words, `SVR`, `NuSVR` and `OneClass` has a deprecated property and `BaseSVC` defines `n_support_`. Then in 1.3 we can remove the property in `SVR`, `NuSVC` and `OneClass`.

thomasjpfan commented on sklearn/svm/_classes.py:
```diff
@@ -1247,6 +1255,15 @@ def __init__(
             random_state=None,
         )
 
+    # TODO: Remove in 1.3
```
```suggestion
    # TODO(1.3): Remove
```

thomasjpfan commented on sklearn/svm/tests/test_svm.py:
```diff
@@ -1425,17 +1424,18 @@ def test_n_support_oneclass_svr():
+    with pytest.warns(FutureWarning):
+        assert not hasattr(clf, "n_support_")
+        clf.fit(X)
+        assert clf.n_support_ == clf.support_vectors_.shape[0]
+        assert clf.n_support_.size == 1
+        assert clf.n_support_ == 3
```
Interesting that we test this. I would be okay with testing `_n_support[0]` directly for these properties.

thomasjpfan commented on sklearn/svm/tests/test_svm.py:
```diff
@@ -1519,3 +1519,13 @@ def test_n_iter_libsvm(estimator, expected_n_iter_type, dataset):
         assert n_iter.shape == (n_classes * (n_classes - 1) // 2,)
+
+
+@pytest.mark.parametrize("Klass", [SVR, NuSVR, OneClassSVM])
```
I'm trying to push for these "TODO(version_number)" so they are easier to grep for.

```suggestion
# TODO(1.3) Remove
@pytest.mark.parametrize("Klass", [SVR, NuSVR, OneClassSVM])
```

Micky774 committed:
Merge branch 'main' into deprecate_in_svr

Micky774 committed:
Update TODO format

Co-authored-by: Thomas J. Fan <thomasjpfan@gmail.com>

Micky774 committed:
Edits to include review comments

- Updated tests to use `n_support_[0]` directly
- Added deprecated property to `SVR, NuSVR, OneClassSVM`
- Removed partial `n_support_` deprecation in parent `BaseLibSVM`
- Removed reliance on `n_support_` in `_validate_for_predict`

Micky774 committed:
Merge branch 'main' into deprecate_in_svr

thomasjpfan commented on sklearn/svm/_base.py:
```diff
@@ -677,14 +680,18 @@ def n_support_(self):
             check_is_fitted(self)
         except NotFittedError:
             raise AttributeError
+        return self._n_support
```
I think it's better to move this into `BaseSVC`. This way when `n_support_` is removed, then we only need to remove the `@property` from the regressors.

thomasjpfan commented on sklearn/svm/_base.py:
```diff
@@ -677,14 +680,18 @@ def n_support_(self):
+    @property
+    def _class_weight(self):
+        """Weights per class"""
+        # Avoid using self.class_weight_ in SVR, NuSVR, OneClass due
+        # due to deprecation
+        return (
+            self.class_weight_
+            if LIBSVM_IMPL.index(self._impl) in (0, 1)
+            else np.empty(0)
+        )
```
I prefer to use `hasattr` here:

```suggestion
        # Class weights are defined for classifiers during
        # fit, otherwise it is empty
        if hasattr(self, "class_weight_"):
            return self.class_weight_
        return np.empty(0)
```

This way we do not couple `LIBSVM_IMPL.index(self._impl) in (0, 1)` with class weights. The logic becomes "If there are class weights, use them, otherwise do not".

thomasjpfan reviewed (commented):
Thank you for the update!

Micky774 committed:
Update sklearn/svm/_base.py

Co-authored-by: Thomas J. Fan <thomasjpfan@gmail.com>

Micky774 committed:
Merge branch 'main' into deprecate_in_svr

Micky774 committed:
Merge branch 'deprecate_in_svr' of https://github.com/Micky774/scikit-learn into deprecate_in_svr

Micky774 committed:
Improved tests and changed handling of `_class_weight`

thomasjpfan commented on sklearn/svm/_classes.py:
```diff
@@ -1362,6 +1392,9 @@ class NuSVR(RegressorMixin, BaseLibSVM):
     n_support_ : ndarray of shape (n_classes,), dtype=int32
         Number of support vectors for each class.
 
+        .. deprecated::
```
The version is included here:

```suggestion
        .. deprecated:: 1.2
```

thomasjpfan commented on sklearn/svm/_classes.py:
```diff
@@ -1362,6 +1392,9 @@ class NuSVR(RegressorMixin, BaseLibSVM):
         Number of support vectors for each class.
 
+        .. deprecated::
+            `n_support_` was deprecated in version 1.1 and will be removed in 1.3.
```
Also all the versions need to be updated.

```suggestion
            `n_support_` was deprecated in version 1.2 and will be removed in 1.4.
```

thomasjpfan commented on doc/whats_new/v1.2.rst:
```diff
@@ -40,3 +40,10 @@ Thanks to everyone who has contributed to the maintenance and improvement of
+:mod:`sklearn.svm`
+..................
+
+- |API| Attributes `n_support_` and `class_weight_` are noe deprecated for
```
```suggestion
- |API| Attributes `n_support_` and `class_weight_` are now deprecated for
```

thomasjpfan commented on sklearn/svm/_classes.py:
```diff
@@ -1550,6 +1605,9 @@ class OneClassSVM(OutlierMixin, BaseLibSVM):
         Number of support vectors for each class.
 
+        .. deprecated::
+            `n_support_` was deprecated in version 1.1 and will be removed in 1.3.
```
I think `n_support_` still makes sense to have in `SVR` and `OneClassSVM` because it is the number of support vectors. In this case, the `n_support_`'s docstring can be updated to say:

> n_support_ : ndarray of shape (1,), dtype=int32

What do you think @amueller?

Micky774 committed:
Corrected version in property deprecation and documentation

Micky774 committed:
Merge branch 'main' into deprecate_in_svr

Micky774 committed:
Update doc/whats_new/v1.2.rst

Co-authored-by: Thomas J. Fan <thomasjpfan@gmail.com>

thomasjpfan commented on sklearn/svm/_base.py:
```diff
@@ -671,20 +674,11 @@ def _get_coef(self):
         return safe_sparse_dot(self._dual_coef_, self.support_vectors_)
 
     @property
-    def n_support_(self):
```
Looking at this again, I think it makes sense to have `n_support_` for regressors and not deprecate it.

WDYT @micky774?

Micky774 replied:
I'm not really a fan of it in the context of `(Nu)SVR` as-is. Its usage in other `SVM` estimators as a `(num_classes, )` tuple seems to imply that the regressors operate with a single class which imo is not quite fair. 

With that being said I would be more friendly to it if we also alter the documentation from
> Number of support vectors for each class.

to
> Number of support vectors.

just to further clarify the distinct semantics regarding the shape of the attribute in regressors.

thomasjpfan replied:
Now that are we not deprecating `n_support_` anymore, we technically do not need to do the `n_support_` refactor. To make this a quicker review for others, we can revert the changes for `n_support_` and keep the `class_weight_` deprecation.

Let's open a small second PR to update `n_support_` documentation.

Micky774 replied:
Opened here: #23353

Micky774 committed:
Merge branch 'main' into deprecate_in_svr

Micky774 committed:
Updated test for version change

Micky774 committed:
Added support for `n_support_`

Micky774 committed:
Updated non-regression test to account for `n_support` changes

Micky774 committed:
Merge branch 'main' into deprecate_in_svr

Micky774 committed:
Reverted changes to `n_support`

Micky774 committed:
Merge branch 'main' into deprecate_in_svr

Micky774 committed:
Merge branch 'main' into deprecate_in_svr

thomasjpfan commented on doc/whats_new/v1.2.rst:
```diff
@@ -108,3 +108,10 @@ Thanks to everyone who has contributed to the maintenance and improvement of
+:mod:`sklearn.svm`
+..................
+
+- |API| Attributes `n_support_` and `class_weight_` are now deprecated for
```
```suggestion
- |API| Attributes `class_weight_` are now deprecated for
```

thomasjpfan commented on sklearn/svm/tests/test_svm.py:
```diff
@@ -1498,3 +1498,16 @@ def test_n_iter_libsvm(estimator, expected_n_iter_type, dataset):
+    with warnings.catch_warnings():
+        warnings.simplefilter("error", FutureWarning)
+        clf.fit(X, Y)
+    attr = "class_weight_"
+    msg = f"Attribute `{attr}` was deprecated in version 1.2 and will be removed in 1.4"
```
Now that there is only one attribute deprecated:

```suggestion
    msg = f"Attribute `class_weight_` was deprecated in version 1.2 and will be removed in 1.4"
```

thomasjpfan commented on sklearn/svm/_base.py:
```diff
@@ -335,7 +332,7 @@ def _dense_fit(self, X, y, sample_weight, solver_type, kernel, random_seed):
             svm_type=solver_type,
             sample_weight=sample_weight,
-            class_weight=self.class_weight_,
+            class_weight=self._class_weight,
```
Do you think this is easier to follow or using `@property`?

```python
class_weight=getattr(self, "_class_weight", np.empty(0))
```

For me, it's about the same.

jjerphan replied:
I agree with @thomasjpfan's suggestions since this simplify the logic not using overridden properties.

Taking your comment in again:
```suggestion
            # Class weights are defined for classifiers during
            # fit, otherwise it is empty.
            class_weight=getattr(self, "_class_weight", np.empty(0)),
```

What do you think?

Micky774 replied:
I'm not sure I quite understand this. The original intention for this approach was to ensure there wouldn't be a `FutureWarning` raised during `fit` or `predict` for `SVR/NuSVR/OneClassSVM` but using `getattr` instead does raise that warning. So then we'd either have to suppress the warning locally, or keep the current form and I prefer the latter.

I may very well just be misunderstanding the suggestion though, so feedback welcome :)

Micky774 replied:
Not sure I understand. The use of `_class_weight` is to ensure `fit` and `predict` don't raise a `FutureWarning`, but using the proposed approach w/ `getattr` still raises it, therefore we'd need to suppress it locally (in 5 places) which seems messy. I may just be misunderstanding the suggestion though 😅

jjerphan commented on sklearn/svm/tests/test_svm.py:
```diff
@@ -1498,3 +1498,16 @@ def test_n_iter_libsvm(estimator, expected_n_iter_type, dataset):
+
+# TODO(1.4): Remove
+@pytest.mark.parametrize("Klass", [SVR, NuSVR, OneClassSVM])
+def test_libsvm_nclass_attrs_deprecated(Klass):
```
Is this more exact?
```suggestion
def test_svm_class_weights_deprecation(Klass):
```

jjerphan reviewed (commented):
Thank you for your this contribution, @Micky774.

Here are a few comments.

Micky774 committed:
Merge branch 'main' into deprecate_in_svr

Micky774 committed:
Updated changelog entry and improved test per reviews

Micky774 committed:
Minor refactor of `_class_weight` property

thomasjpfan commented on sklearn/svm/_base.py:
```diff
@@ -335,7 +332,7 @@ def _dense_fit(self, X, y, sample_weight, solver_type, kernel, random_seed):
             svm_type=solver_type,
             sample_weight=sample_weight,
-            class_weight=self.class_weight_,
+            class_weight=getattr(self, "_class_weight", np.empty(0)),
```
I think there is a `TODO(1.4)` here to change `_class_weight` to `class_weight_`?

Same comment for the other locations.

thomasjpfan replied:
I think `_class_weight` was used to avoid the depreciation warning?

Micky774 replied:
Indeed, using `getattr(self, "class_weight_", np.empty(0))` results in a `FutureWarning`.

thomasjpfan reviewed (approved):
Minor comment, otherwise LGTM

Micky774 committed:
Added TODO comments

Micky774 committed:
Merge branch 'main' into deprecate_in_svr

ogrisel reviewed (approved):
It took me time to get it but it looks correct and the tests pass without raising the warning when it should not be raised.

The trick to look at the `_validate_targets` method that is overridden in the classifier base class `BaseSVC` that defines the public `class_weight_` only for classifiers.
